driver-core/src/main/com/mongodb/client/model/expressions/Expression.java

```diff
@@ -111,5 +111,7 @@ public interface Expression {
     <T extends Expression> ArrayExpression<T> isArrayOr(ArrayExpression<? extends T> other);
     <T extends DocumentExpression> T isDocumentOr(T other);
 
+    <T extends Expression> MapExpression<T> isMapOr(MapExpression<T> other);
```

Comment:
To me this seems fine and unrelated to the sufferings of `isArrayOr`. The API makes a strong and encapsulating distinction between the types of the API, and the implementing "structures". The same distinction can be found with Java's `LinkedList` and `ArrayList` structures, and the higher-level types `Queue` and `List`. Consider this pseudocode:

```
Queue queue = new LinkedList(1, 2, 3);
List list = new ArrayList(1, 2, 3);
queue.equals(list); // true
list = (List) queue;
```
Our API is different in important ways (for example, we never cast), but the `isSomethingOr` serve the same purpose. For example, `myInteger.isNumberOr(...)` will return the original integer value. Of course, integer inherits from number, but the fact that map/doc do not (presently) inherit does not bother me, because the behaviour of Queue and List (which I see as expected and correct) does not bother me. That is, types can share values, and do so without an explicit hierarchical relationship.

Reply:
> we never cast

We do cast, what we don't do is we don't require a user to cast explicitly. We cast and hope for the best, assuming that the user will only do these implicit casts when the casts are actually correct.

> types can share values, and do so without an explicit hierarchical relationship

If we think of a type as "a set of values", then your statement is problematic when two sets of values are the same (as they are in our case). And that's how I was thinking about types. However, if we say that a type is a combination of a set of values and what we can do with each value (we may be allowed to do different things with different values of the same type, e.g., we can use most `int` values as a denominator but we can't use 0 as a denominator), then your statement does not have a problem.